Make make_dual redispatch

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68631
* #68629
* __->__ #68630
* #68628

Constraints:
1) (functorch) if all the inputs to an op have requires_grad=False and don't have tangents, then their VariableType
    kernel should be a no-op i.e., behave like a redispatch. This is due to functorch's DynamicLayerStack
   having the autograd key by default (which is so that transformations like vmap) still work with autograd
2) (inference mode) inference tensors in inference mode will call straight into the kernel, we should still do something sensible   
    inside even if we normally wouldn't redispatch into it.
3) ~Should support potential application of interposition below autograd: `nn.Parameter` is a example of subclassing where the subclass  
    is not preserved when an operation is performed. There is an exception though: we want calling `make_dual` on a 
    `nn.Parameter` to preserve its parameterness.~
4) Should avoid calls to shallow_copy_and_detach to avoid spurious calls into `__python_dispatch__`. 

This PR:
- does not redispatch to `make_dual` from its `ADInplaceOrView` kernel to satisfy (1)
- calls into `alias` from the kernel in the native namespace so that behavior is consistent with other views in inference mode to satisfy (2)
- discussion of (3). We still wouldn't be able to directly override `make_dual` below autograd. In this PR, instead of not redispatching at all, we choose to redispatch into `at::alias` so that one can override `make_dual`. The side effect is that one would not be able to distinguish calls between the two, which can be problematic (though a straightforward but hacky solution would be to create a new `at::alias_for_make_dual` that would allow users to distinguish) the two. This isn't ideal but seems to be the simplest way to satisfy (3). We don't pursue that hacky solution here.
- (4) is satisfied because we remove calls to `shallow_copy_and_detach`

<details>
<summary> A potentially less hacky but more involved solution? (WIP) </summary>

Realizing that make_dual is more like requires_grad, perhaps it shouldn't be autograd explicit? Make make_dual a composite or python-only construct. i.e., it would be a view on the primal followed by something to the effect of primal.set_fw_grad(tangent).
         
Additional constraints:
5) make_dual needs to be backward-differentiable (I can't think of any applications yet becuase
   technically as a high-order function, jvp's input is the tangent only, "detach" is not applied on
   the tangent, so one would still be able to propagate gradients through it).
6) set_fw_grad needs to raise an error if there is a layout mismatch and base is a forward-differnentiable view

Possible plan
- (6) implies that a plain view would not suffice. We need a `detach`-like operation to ensure that set_fw_grad 
  knows the view is not forward differentiable.
- (5) implies that is this (new) `detach` would need to be backward differentiable (API TBD).
- (3) is no longer relevant because make_dual is no longer autograd explicit, but perhaps this new detach should behave like the current one? There is a lot of logic to replicate for detach, so this may be hard.
- (1) is satisfied if we use current detach logic, i.e., , and (4) is trivial.

I'm not convinced that this is the right solution either, because in the end does (3) still work?

 </details>

Differential Revision: [D32899679](https://our.internmc.facebook.com/intern/diff/D32899679)

soulitzer committed:
Make make_dual properly dispatch to native function

[ghstack-poisoned]

soulitzer committed:
Update on "[WONT MERGE] Make make_dual properly dispatch to native function"


There is a constraint in functorch that if all the inputs to an op have requires_grad=False and don't have tangents, then their VariableType kernel must be a no-op i.e., behave like a redispatch. This is due to functorch's DynamicLayerStack having the autograd key by default (which is so that transformations like vmap) still work with autograd in core.

I think it is ok if we don't end up supporting subclassing to override make_dual because there isn't something analogous for requires_grad either.

Making sure inference mode has the correct behavior can be done using alternative means

[ghstack-poisoned]

soulitzer committed:
Update on "[WONT MERGE] Make make_dual properly dispatch to native function"



TLDR: In its current state, we satisfy (1), (2), (3), and (4) is trivial. Ideally we'd also support (5) and (6), but (5) naturally clashes with (3). A method to reconcile (3) and (5) is described in appendix (see very bottom), but it seems tricky. We propose two ways forward option 1 and 2 and discuss their tradeoffs.
         
Possible constraints:
1) make_dual needs to be backward-differentiable (I can't think of any applications yet becuase
   technically as a high-order function, jvp's input is the tangent only, "detach" is not applied on
   the tangent, so one would still be able to propagate gradients through it).
2) set_fw_grad needs to raise an error if there is a layout mismatch and base is a forward-differnentiable view
3) (functorch) if all the inputs to an op have requires_grad=False and don't have tangents, then their VariableType
   kernel must be a no-op i.e., behave like a redispatch. This is due to functorch's DynamicLayerStack
   having the autograd key by default (which is so that transformations like vmap) still work with autograd
   in core. (I wonder if the subclassing implementation of functorch will solve this.)
4) Inference tensors in inference mode should not error.
5) python dispatch should be able to interpose (if make_dual is a non-forward-differentiable view),
   because this case is analagous to detach. More generally, any detach-like construct should 
   allow python dispatch to interpose.
6) In its current state, make_dual triggers python dispatch per shallow_copy_and_detach
   in which the input is a dual tensor. An additional constraint would be the clean up of this
   spurious dispatch, by... excluding the Python key?

Option 1: Make make_dual a composite or python-only construct. i.e., it would be a view on the 
primal followed by something to the effect of primal.set_fw_grad(tangent).
- (2) implies that a plain view would not suffice. We need a `detach` to ensure that set_fw_grad 
  knows the view is not forward differentiable.
- (1) implies that is this `detach` would need to be backward differentiable (API TBD).
- (5) implies that we would need to replicate the current detach logic for this new detach.
- (3) is satisfied if we use current detach logic, and (4) is trivial.

Option 2: keep make_dual as CompositeExplicitAutograd (what it currently is), and then attack (5)
- (1), (2), (3) are satisfied. (4) is trivial.
- (5) implies that we would need to replicate the current detach logic for make_dual

Tradeoffs: To satisify constraint (5) both option 1 and option 2 requires the replication of current detach
logic. Option 1 seems slightly easier in terms of satisfying (5) because we'd just need a new
argument on detach. Option 1 makes (6) irrelevant as well, because make_dual wouldn't
need to be a primitive wrt Python dispatch anymore.

Appendix: why detach is able to work even though it doesn't redispatch?
- Detach calls into shallow_copy_and_detach, in which if the Python key exists on the tensor and isn't
  excluded we call into a special detach "virtual" method of the PyInterpreter "concrete_detach_fn",
  which calls straight into the `__python_dispatch__` implementation. 
- Because _fw_primal and make_dual call into shallow_copy_and_detach, in LoggingTensor we actually
  go into `__python_dispatch__` on behalf of detach as well. Side note: because `__repr__` of a dual tensor
  calls into unpack_dual which calls into _fw_primal, in its current state we actually run into the
  "not implemented" error.
  

[ghstack-poisoned]

soulitzer committed:
Update on "[WONT MERGE] Make make_dual properly dispatch to native function"



TLDR: In its current state, we satisfy (1), (2), (3), and (4) is trivial. Ideally we'd also support (5) and (6), but (5) naturally clashes with (3). A method to reconcile (3) and (5) is described in appendix (see very bottom), but it seems tricky. We propose two ways forward option 1 and 2 and discuss their tradeoffs.
         
Possible constraints:
1) make_dual needs to be backward-differentiable (I can't think of any applications yet becuase
   technically as a high-order function, jvp's input is the tangent only, "detach" is not applied on
   the tangent, so one would still be able to propagate gradients through it).
2) set_fw_grad needs to raise an error if there is a layout mismatch and base is a forward-differnentiable view
3) (functorch) if all the inputs to an op have requires_grad=False and don't have tangents, then their VariableType
   kernel must be a no-op i.e., behave like a redispatch. This is due to functorch's DynamicLayerStack
   having the autograd key by default (which is so that transformations like vmap) still work with autograd
   in core. (I wonder if the subclassing implementation of functorch will solve this.)
4) Inference tensors in inference mode should not error.
5) python dispatch should be able to interpose (if make_dual is a non-forward-differentiable view),
   because this case is analagous to detach. More generally, any detach-like construct should 
   allow python dispatch to interpose.
6) In its current state, make_dual triggers python dispatch per shallow_copy_and_detach
   in which the input is a dual tensor. An additional constraint would be the clean up of this
   spurious dispatch, by... excluding the Python key?

Option 1: Make make_dual a composite or python-only construct. i.e., it would be a view on the 
primal followed by something to the effect of primal.set_fw_grad(tangent).
- (2) implies that a plain view would not suffice. We need a `detach` to ensure that set_fw_grad 
  knows the view is not forward differentiable.
- (1) implies that is this `detach` would need to be backward differentiable (API TBD).
- (5) implies that we would need to replicate the current detach logic for this new detach.
- (3) is satisfied if we use current detach logic, and (4) is trivial.

Option 2: keep make_dual as CompositeExplicitAutograd (what it currently is), and then attack (5)
- (1), (2), (3) are satisfied. (4) is trivial.
- (5) implies that we would need to replicate the current detach logic for make_dual

Tradeoffs: To satisify constraint (5) both option 1 and option 2 requires the replication of current detach
logic. Option 1 seems slightly easier in terms of satisfying (5) because we'd just need a new
argument on detach. Option 1 makes (6) irrelevant as well, because make_dual wouldn't
need to be a primitive wrt Python dispatch anymore.

Appendix: why detach is able to work even though it doesn't redispatch?
- Detach calls into shallow_copy_and_detach, in which if the Python key exists on the tensor and isn't
  excluded we call into a special detach "virtual" method of the PyInterpreter "concrete_detach_fn",
  which calls straight into the `__python_dispatch__` implementation. 
- Because _fw_primal and make_dual call into shallow_copy_and_detach, in LoggingTensor we actually
  go into `__python_dispatch__` on behalf of detach as well. Side note: because `__repr__` of a dual tensor
  calls into unpack_dual which calls into _fw_primal, in its current state we actually run into the
  "not implemented" error.
  

[ghstack-poisoned]

soulitzer committed:
Update on "[WONT MERGE] Make make_dual properly dispatch to native function"


Constraints:
1) (functorch) if all the inputs to an op have requires_grad=False and don't have tangents, then their VariableType
    kernel should be a no-op i.e., behave like a redispatch. This is due to functorch's DynamicLayerStack
   having the autograd key by default (which is so that transformations like vmap) still work with autograd
2) (inference mode) inference tensors in inference mode will call straight into the kernel, we should still do something sensible   
    inside even if we normally wouldn't redispatch into it.
3) Should support potential application of interposition below autograd: `nn.Parameter` is a example of subclassing where the subclass  
    is not preserved when an operation is performed. There is an exception though: we want calling `make_dual` on a 
    `nn.Parameter` to preserve its parameterness.
4) Should avoid calls to shallow_copy_and_detach to avoid spurious calls into `__python_dispatch__`. 

This PR:
- does not redispatch to `make_dual` from its `ADInplaceOrView` kernel to satisfy (1)
- calls into `alias` from the kernel in the native namespace so that behavior is consistent with other views in inference mode to satisfy (2)
- partial satisfaction of (3). We still wouldn't be able to directly override `make_dual` below autograd. In this PR, instead of not redispatching at all, we choose to redispatch into `at::alias` so that one can override `make_dual`. The side effect is that one would not be able to distinguish calls between the two, which can be problematic (though a straightforward but hacky solution would be to create a new `at::alias_for_make_dual` that would allow users to distinguish) the two. This isn't ideal but seems to be the simplest way to satisfy (3).
- (4) is satisfied because we remove calls to `shallow_copy_and_detach`

<details>
<summary> A potentially less hacky solution? (WIP) </summary>

Realizing that make_dual is more like requires_grad, perhaps it shouldn't be autograd explicit? Make make_dual a composite or python-only construct. i.e., it would be a view on the primal followed by something to the effect of primal.set_fw_grad(tangent).
         
Additional constraints:
5) make_dual needs to be backward-differentiable (I can't think of any applications yet becuase
   technically as a high-order function, jvp's input is the tangent only, "detach" is not applied on
   the tangent, so one would still be able to propagate gradients through it).
6) set_fw_grad needs to raise an error if there is a layout mismatch and base is a forward-differnentiable view

Possible plan
- (6) implies that a plain view would not suffice. We need a `detach`-like operation to ensure that set_fw_grad 
  knows the view is not forward differentiable.
- (5) implies that is this (new) `detach` would need to be backward differentiable (API TBD).
- (3) is no longer relevant because make_dual is no longer autograd explicit, but perhaps this new detach should behave like the current one? There is a lot of logic to replicate for detach, so this may be hard.
- (1) is satisfied if we use current detach logic, i.e., , and (4) is trivial.

I'm not convinced that this is the right solution either, because in the end does (3) still work?

 </details>

[ghstack-poisoned]

soulitzer commented:
From offline discussion: add check that verifies that going into native function can only be done in inference mode.

albanD commented on aten/src/ATen/native/AutogradComposite.cpp:
```diff
@@ -4,12 +4,8 @@
 namespace at {
 namespace native {
 
-// _make_dual is actually implemented in VariableTypeManual.cpp, this is just a stub
```
I think we could leave a comment here saying that this should only be called in inference mode.

albanD reviewed (approved):
Are you sure about 3. ?
Or was it dropped and the description wasn't updated?

This is a good fix for the inference. We still need to make a choice if we want make_dual/unpack_dual/detach to be composite implicit or not.

soulitzer commented:
> Are you sure about 3. ?

Ehhh, probably a good idea to cross it out from this list, since we don't really address it here, but good to support eventually I think.

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32899679).

soulitzer committed:
Update on "Make make_dual redispatch"


Constraints:
1) (functorch) if all the inputs to an op have requires_grad=False and don't have tangents, then their VariableType
    kernel should be a no-op i.e., behave like a redispatch. This is due to functorch's DynamicLayerStack
   having the autograd key by default (which is so that transformations like vmap) still work with autograd
2) (inference mode) inference tensors in inference mode will call straight into the kernel, we should still do something sensible   
    inside even if we normally wouldn't redispatch into it.
3) ~Should support potential application of interposition below autograd: `nn.Parameter` is a example of subclassing where the subclass  
    is not preserved when an operation is performed. There is an exception though: we want calling `make_dual` on a 
    `nn.Parameter` to preserve its parameterness.~
4) Should avoid calls to shallow_copy_and_detach to avoid spurious calls into `__python_dispatch__`. 

This PR:
- does not redispatch to `make_dual` from its `ADInplaceOrView` kernel to satisfy (1)
- calls into `alias` from the kernel in the native namespace so that behavior is consistent with other views in inference mode to satisfy (2)
- discussion of (3). We still wouldn't be able to directly override `make_dual` below autograd. In this PR, instead of not redispatching at all, we choose to redispatch into `at::alias` so that one can override `make_dual`. The side effect is that one would not be able to distinguish calls between the two, which can be problematic (though a straightforward but hacky solution would be to create a new `at::alias_for_make_dual` that would allow users to distinguish) the two. This isn't ideal but seems to be the simplest way to satisfy (3). We don't pursue that hacky solution here.
- (4) is satisfied because we remove calls to `shallow_copy_and_detach`

<details>
<summary> A potentially less hacky but more involved solution? (WIP) </summary>

Realizing that make_dual is more like requires_grad, perhaps it shouldn't be autograd explicit? Make make_dual a composite or python-only construct. i.e., it would be a view on the primal followed by something to the effect of primal.set_fw_grad(tangent).
         
Additional constraints:
5) make_dual needs to be backward-differentiable (I can't think of any applications yet becuase
   technically as a high-order function, jvp's input is the tangent only, "detach" is not applied on
   the tangent, so one would still be able to propagate gradients through it).
6) set_fw_grad needs to raise an error if there is a layout mismatch and base is a forward-differnentiable view

Possible plan
- (6) implies that a plain view would not suffice. We need a `detach`-like operation to ensure that set_fw_grad 
  knows the view is not forward differentiable.
- (5) implies that is this (new) `detach` would need to be backward differentiable (API TBD).
- (3) is no longer relevant because make_dual is no longer autograd explicit, but perhaps this new detach should behave like the current one? There is a lot of logic to replicate for detach, so this may be hard.
- (1) is satisfied if we use current detach logic, i.e., , and (4) is trivial.

I'm not convinced that this is the right solution either, because in the end does (3) still work?

 </details>

Differential Revision: [D32899679](https://our.internmc.facebook.com/intern/diff/D32899679)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32899679).

soulitzer committed:
Update on "Make make_dual redispatch"


Constraints:
1) (functorch) if all the inputs to an op have requires_grad=False and don't have tangents, then their VariableType
    kernel should be a no-op i.e., behave like a redispatch. This is due to functorch's DynamicLayerStack
   having the autograd key by default (which is so that transformations like vmap) still work with autograd
2) (inference mode) inference tensors in inference mode will call straight into the kernel, we should still do something sensible   
    inside even if we normally wouldn't redispatch into it.
3) ~Should support potential application of interposition below autograd: `nn.Parameter` is a example of subclassing where the subclass  
    is not preserved when an operation is performed. There is an exception though: we want calling `make_dual` on a 
    `nn.Parameter` to preserve its parameterness.~
4) Should avoid calls to shallow_copy_and_detach to avoid spurious calls into `__python_dispatch__`. 

This PR:
- does not redispatch to `make_dual` from its `ADInplaceOrView` kernel to satisfy (1)
- calls into `alias` from the kernel in the native namespace so that behavior is consistent with other views in inference mode to satisfy (2)
- discussion of (3). We still wouldn't be able to directly override `make_dual` below autograd. In this PR, instead of not redispatching at all, we choose to redispatch into `at::alias` so that one can override `make_dual`. The side effect is that one would not be able to distinguish calls between the two, which can be problematic (though a straightforward but hacky solution would be to create a new `at::alias_for_make_dual` that would allow users to distinguish) the two. This isn't ideal but seems to be the simplest way to satisfy (3). We don't pursue that hacky solution here.
- (4) is satisfied because we remove calls to `shallow_copy_and_detach`

<details>
<summary> A potentially less hacky but more involved solution? (WIP) </summary>

Realizing that make_dual is more like requires_grad, perhaps it shouldn't be autograd explicit? Make make_dual a composite or python-only construct. i.e., it would be a view on the primal followed by something to the effect of primal.set_fw_grad(tangent).
         
Additional constraints:
5) make_dual needs to be backward-differentiable (I can't think of any applications yet becuase
   technically as a high-order function, jvp's input is the tangent only, "detach" is not applied on
   the tangent, so one would still be able to propagate gradients through it).
6) set_fw_grad needs to raise an error if there is a layout mismatch and base is a forward-differnentiable view

Possible plan
- (6) implies that a plain view would not suffice. We need a `detach`-like operation to ensure that set_fw_grad 
  knows the view is not forward differentiable.
- (5) implies that is this (new) `detach` would need to be backward differentiable (API TBD).
- (3) is no longer relevant because make_dual is no longer autograd explicit, but perhaps this new detach should behave like the current one? There is a lot of logic to replicate for detach, so this may be hard.
- (1) is satisfied if we use current detach logic, i.e., , and (4) is trivial.

I'm not convinced that this is the right solution either, because in the end does (3) still work?

 </details>

Differential Revision: [D32899679](https://our.internmc.facebook.com/intern/diff/D32899679)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32899679).
